x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/import/calculate_rule_source_from_asset.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RuleResponse, RuleSource } from '../../../../../../common/api/detection_engine';
+import type { PrebuiltRuleAsset } from '../../../prebuilt_rules';
+import { calculateIsCustomized } from '../detection_rules_client/mergers/rule_source/calculate_is_customized';
+
+/**
+ * Calculates rule_source for a rule based on two pieces of information:
+ * 1. The prebuilt rule asset that matches the specified rule_id and version
+ * 2. Whether a prebuilt rule with the specified rule_id is currently installed
+ *
+ * @param rule The rule for which rule_source is being calculated
+ * @param assetWithMatchingVersion The prebuilt rule asset that matches the specified rule_id and version
+ * @param ruleIdExists Whether a prebuilt rule with the specified rule_id is currently installed
+ *
+ * @returns The calculated rule_source
+ */
+export const calculateRuleSourceFromAsset = ({
+  rule,
+  assetWithMatchingVersion,
+  ruleIdExists,
+}: {
+  rule: RuleResponse;
+  assetWithMatchingVersion: PrebuiltRuleAsset | undefined;
+  ruleIdExists: boolean;
+}): RuleSource => {
+  if (!ruleIdExists) {
+    return {
+      type: 'internal',
+    };
+  }
+
+  if (assetWithMatchingVersion == null) {
+    return {
+      type: 'external',
+      is_customized: true,
+    };
+  }
+
+  const isCustomized = calculateIsCustomized(assetWithMatchingVersion, rule);
```

Comment:
I agree with the general sentiment; I had intended for whatever logic was introduced as part of this PR to consolidate the existing implementations. However, the complexity/ambiguity of the existing code lead me to develop the `RuleSourceImporter` independently as a special case for importing. I think that it could be reused/renamed as `BulkRuleSourceCalculator` or something to that effect.

In my mind there are four similar (but separate) code paths that we're talking about here:

- `calculateIsCustomized`
- `calculateRuleSource`
- `calculateRuleSourceForImport`/`calculateRuleSourceFromAsset`
- `normalizeRuleSource`

I had intended to at least rewrite `calculateRuleSource` in terms of `calculateRuleSourceFromAsset`, but didn't end up doing so once I realized that they have different semantics (`calculateRuleSource` uses/respects the rule's `immutable` value, while the `FromAsset` variant does not).